silence/silence_test.go

```diff
@@ -109,6 +109,119 @@ func TestSilencesGC(t *testing.T) {
 	require.Equal(t, want, s.st)
 }
 
+func TestSilenceGCOverTime(t *testing.T) {
+	type silenceEntry struct {
+		s                    *pb.Silence
+		expectPresentAfterGc bool
+	}
+	type testCase struct {
+		initialState    []silenceEntry
+		updates         []silenceEntry
+		expectedGCCount int
+	}
+
+	c := clock.NewMock()
+	now := c.Now().UTC()
+
+	newSilence := func(id string, exp time.Time) *pb.Silence {
+		return &pb.Silence{
+			Id:       id,
+			StartsAt: now.Add(-time.Second),
+			EndsAt:   exp,
+			Matchers: []*pb.Matcher{
+				{Name: "foo", Type: pb.Matcher_REGEXP, Pattern: "bar"},
+			}}
+	}
+
+	cases := map[string]testCase{
+		"gc does not clean active silences": {
+			initialState: []silenceEntry{
+				{s: newSilence("1", now), expectPresentAfterGc: false},
+				{s: newSilence("2", now.Add(-time.Second)), expectPresentAfterGc: false},
+				{s: newSilence("3", now.Add(time.Second)), expectPresentAfterGc: true},
+			},
+		},
+		"silences added with Set are handled correctly": {
+			initialState: []silenceEntry{
+				{s: newSilence("1", now), expectPresentAfterGc: false},
+			},
+			updates: []silenceEntry{
+				{s: newSilence("", now.Add(time.Second)), expectPresentAfterGc: true},
+				{s: newSilence("", now.Add(-time.Second)), expectPresentAfterGc: false},
+			},
+		},
+		"silence update does not leak state": {
+			initialState: []silenceEntry{
+				{s: newSilence("1", now), expectPresentAfterGc: false},
+			},
+			updates: []silenceEntry{
+				{s: newSilence("1", now.Add(time.Second)), expectPresentAfterGc: true},
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			silences, err := New(Options{})
+			silClock := clock.NewMock()
+			silences.clock = silClock
+			require.NoError(t, err)
+
+			expectedRemaining := []string{}
+			expectedGCCount := 0
+			for _, sil := range tc.initialState {
+				silences.st[sil.s.Id] = &pb.MeshSilence{
+					Silence:   sil.s,
+					ExpiresAt: sil.s.EndsAt,
+				}
+				if sil.expectPresentAfterGc {
+					expectedRemaining = append(expectedRemaining, sil.s.Id)
+				} else {
+					expectedGCCount += 1
+				}
+				// simulate this silences being seen in a query
+				silences.mc.Get(silences.st[sil.s.Id].Silence)
+			}
+			silClock.Add(-time.Second)
```

Comment:
I'm not sure I understand the significance of rewinding the clock, although I can see that if I comment this out the test fails. How does rewinding the clock help us test GC behavior?

Reply:
Thanks for point this out! I did this because adding a silence with `Set` that expires before the clock's `now` is a no-op. Typically, this case is handled before we reach `Set`. However, rewinding the clock is a really unclear way to handle this.

Instead, I should've started with the clock 2 seconds behind and then incremented the clock forward by one second after the `initialState` is applied and then again after `updates` are applied. That has the exact same behavior, but is much more expressive to the reader.